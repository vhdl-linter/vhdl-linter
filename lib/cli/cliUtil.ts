import { readdirSync } from "fs";
import { cwd } from "process";
import { DiagnosticSeverity } from "vscode-languageserver";
import { OIRange } from "../parser/objects";
import { joinURL } from "../projectParser";
import { OIDiagnostic } from "../vhdlLinter";

export interface MessageWrapper {
  file: string,
  messages: (OIDiagnostic | { message: string })[]
}
function isOIDiagnostic(obj: unknown): obj is OIDiagnostic {
  if ((obj as OIDiagnostic).range instanceof OIRange) {
    return true;
  }
  return false;
}
export function readDirPath(path: URL) {
  return readdirSync(path).map(file => joinURL(path, file));
}
function getMessageColor(message: OIDiagnostic | { message: string }) {
  if (isOIDiagnostic(message) && message.severity === DiagnosticSeverity.Error) {
    return '\u001b[31m';
  } else if (isOIDiagnostic(message) && message.severity === DiagnosticSeverity.Warning) {
    return '\u001b[33m';
  }
  return '\u001b[34m';
}
export function prettyPrintMessages(messages: MessageWrapper[]) {
  return messages.map(message => {
    const filename = message.file.replace(cwd(), '');
    return message.messages.slice(0, 5).map((innerMessage) => {
      const messageText = `${getMessageColor(innerMessage)}${innerMessage.message}\u001b[0m`;
      if (isOIDiagnostic(innerMessage)) {
        return `${filename}:${innerMessage.range.start.line + 1} (r: ${innerMessage.range.start.line}:${innerMessage.range.start.character} - ${innerMessage.range.end.line}:${innerMessage.range.end.character})\n  ${messageText}`; // lines are 0 based in OI
      }
      return `${filename}\n  ${messageText}`;
    }).join('\n') + (message.messages.length > 5 ? `\n\u001b[31m ... and ${message.messages.length - 5} more\u001b[0m` : '');
  }).join('\n');
}
