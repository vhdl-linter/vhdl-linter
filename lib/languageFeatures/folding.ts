import { FoldingRange, FoldingRangeKind } from 'vscode-languageserver';
import * as I from '../parser/interfaces';
import * as O from '../parser/objects';
import { VhdlLinter } from '../vhdlLinter';
export function foldingHandler(linter: VhdlLinter): FoldingRange[] {
  const result: FoldingRange[] = [];
  for (const obj of linter.file.objectList) {
    if (I.implementsIHasDeclarations(obj) && obj.declarationsRange !== undefined) {
      result.push(FoldingRange.create(obj.declarationsRange.start.line, obj.declarationsRange.end.line - 1));
    }
    if (I.implementsIHasStatements(obj)) {
      result.push(FoldingRange.create(obj.statementsRange.start.line, obj.statementsRange.end.line - 1));
    }
    if (obj instanceof O.OInstantiation || obj instanceof O.OAssociationList || obj instanceof O.OCaseGenerate || obj instanceof O.OCase || obj instanceof O.OConfiguration ||
      (obj instanceof O.OType && (obj.protected || obj.protectedBody || obj instanceof O.ORecord))) {
      result.push(FoldingRange.create(obj.range.start.line, obj.range.end.line));
    }
    if (I.implementsIHasUseClause(obj) && I.implementsIHasLibraries(obj)) {
      const startLine = [...obj.useClauses, ...obj.libraries].reduce((prev, curr) => Math.min(prev, curr.range.start.line), Number.POSITIVE_INFINITY);
      const endLine = [...obj.useClauses, ...obj.libraries].reduce((prev, curr) => Math.max(prev, curr.range.end.line), 0);
      if (startLine > 0 || endLine > 0) {
        result.push(FoldingRange.create(startLine, endLine, undefined, undefined, FoldingRangeKind.Imports));
      }
    }
    if (I.implementsIHasPorts(obj) && obj.portRange !== undefined) {
      result.push(FoldingRange.create(obj.portRange.start.line, obj.portRange.end.line));
    }
    if (I.implementsIHasGenerics(obj) && obj.genericRange !== undefined) {
      result.push(FoldingRange.create(obj.genericRange.start.line, obj.genericRange.end.line));
    }
  }
  return result;
}
