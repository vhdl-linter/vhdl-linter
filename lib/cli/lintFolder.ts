import { lstatSync } from "fs";
import { ProjectParser, vhdlGlob } from "../projectParser";
import { VhdlLinter } from "../vhdlLinter";
import { readFileSyncNorm } from "./readFileSyncNorm";
import { MessageWrapper, prettyPrintMessages, readDirPath } from "./cliUtil";
import { OIRange } from "../parser/objects";
import { minimatch } from "minimatch";
import { basename, dirname } from "path";
import { fileURLToPath } from "url";
import { execSync } from "child_process";
import { platform } from "process";

let gitExists = false;
if (platform === 'win32') {
  try {
    execSync(`git --version`);
    gitExists = true;
  } catch (e) {
    gitExists = false;
  }
} else {
  try {
    execSync(`command -v git`);
    gitExists = true;
  } catch (e) {
    gitExists = false;
  }
}
function isGitIgnored(path: string) {
  if (path.endsWith('.git')) {
    return true;
  }
  if (gitExists) {
    // git check-ignore will fail if the file is *NOT* ignored or when we are not in a git repository
    try {
      execSync(`git check-ignore ${path}`, {cwd: dirname(path)});
      // is a git repo and the file is ignored...
      return true;
    } catch (e) {
      return false;
    }
  } else {
    return false;
  }
}
function ignore(excludePatterns: string[], relativePath: string, absolutePath: string) {
  return excludePatterns.some(glob => minimatch(relativePath, glob)) || isGitIgnored(absolutePath);
}

export async function lintFolder(path: URL, errorExpected: boolean, printMessages: boolean, excludePatterns: string[], projectParser: ProjectParser): Promise<MessageWrapper[]> {
  const result: MessageWrapper[] = [];
  const rootPath = fileURLToPath(projectParser.workspaces[0]!);
  for (const subPath of readDirPath(path)) {
    // in this case we only ever have one workspace
    const relativePath = fileURLToPath(subPath).replace(rootPath, '').substring(1);
    // ignore excluded folders & files
    if (lstatSync(subPath).isDirectory()) {
      if (ignore(excludePatterns, relativePath, fileURLToPath(subPath))) {
        continue;
      }
      result.push(...await lintFolder(subPath, errorExpected, printMessages, excludePatterns, projectParser));
    } else if (minimatch(basename(relativePath), vhdlGlob)) {
      if (ignore(excludePatterns, relativePath, fileURLToPath(subPath))) {
        continue;
      }
      const text = readFileSyncNorm(subPath, { encoding: 'utf8' });
      const vhdlLinter = new VhdlLinter(subPath, text, projectParser, await projectParser.getDocumentSettings(subPath));
      if (vhdlLinter.parsedSuccessfully) {
        await vhdlLinter.checkAll();
      }
      if (errorExpected === false) {
        if (vhdlLinter.messages.length > 0) {
          const newMessage = {
            file: fileURLToPath(subPath),
            messages: vhdlLinter.messages
          };
          result.push(newMessage);
          if (printMessages) {
            console.log(prettyPrintMessages(rootPath, [newMessage]));
          }
        }
      } else {
        if (vhdlLinter.messages.length !== 1) {
          const newMessage: MessageWrapper = {
            file: fileURLToPath(subPath),
            messages: [
              ...vhdlLinter.messages,
              {
                message: `One message expected found ${vhdlLinter.messages.length}`,
                range: new OIRange(vhdlLinter.file, 0, 0)
              }]
          };
          result.push(newMessage);
          if (printMessages) {
            console.log(prettyPrintMessages(rootPath, [newMessage]));
          }
        }
      }
    }
  }

  return result;
}