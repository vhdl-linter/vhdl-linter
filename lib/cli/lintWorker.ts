import { lstatSync } from "fs";
import { parentPort, workerData } from "worker_threads";
import { ProjectParser } from "../projectParser";
import { VhdlLinter } from "../vhdlLinter";
import { readFileSyncNorm } from "./readFileSyncNorm";
import { MessageWrapper, getCodeClimate, prettyPrintMessages, readDirPath } from "./cliUtil";
import { OIRange } from "../parser/objects";
async function run_test(path: URL, errorExpected: boolean, printMessages: boolean, projectParser?: ProjectParser): Promise<MessageWrapper[]> {
  const result: MessageWrapper[] = [];
  if (!projectParser) {
    projectParser = await ProjectParser.create([path]);
  }
  for (const subPath of readDirPath(path)) {
    if (lstatSync(subPath).isDirectory()) {
      result.push(...await run_test(subPath, errorExpected, printMessages, projectParser));
    } else if (subPath.pathname.match(/\.vhdl?$/i)) {
      const text = readFileSyncNorm(subPath, { encoding: 'utf8' });
      const vhdlLinter = new VhdlLinter(subPath, text, projectParser, await projectParser.getDocumentSettings(subPath));
      if (vhdlLinter.parsedSuccessfully) {
        await vhdlLinter.checkAll();
      }
      if (errorExpected === false) {
        if (vhdlLinter.messages.length > 0) {
          const newMessage = {
            file: subPath.pathname,
            messages: vhdlLinter.messages
          };
          result.push(newMessage);
          if (printMessages) {
            console.log(prettyPrintMessages([newMessage]));
          }
        }
      } else {
        if (vhdlLinter.messages.length !== 1) {
          const newMessage: MessageWrapper = {
            file: subPath.pathname,
            messages: [
              ...vhdlLinter.messages,
              {
                message: `One message expected found ${vhdlLinter.messages.length}`,
                range: new OIRange(vhdlLinter.file, 0, 0)
              }]
          };
          result.push(newMessage);
          if (printMessages) {
            console.log(prettyPrintMessages([newMessage]));
          }
        }
      }
    }
  }

  return result;
}
(async () => {
  const { path, errorExpected, outputCodeClimate } = workerData as { path: string, errorExpected: boolean, outputCodeClimate: boolean };
  const messages = await run_test(new URL(path), errorExpected, outputCodeClimate === false);
  if (outputCodeClimate === true) {
    parentPort?.postMessage(getCodeClimate(messages));
  } else {
    parentPort?.postMessage(messages);
  }
})().catch(err => console.error(err));