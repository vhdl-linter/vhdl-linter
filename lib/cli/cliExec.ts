import { Command, InvalidArgumentError } from '@commander-js/extra-typings';
import { existsSync, lstatSync } from 'fs';
import { isAbsolute, join } from 'path';
import { cwd } from 'process';
import { fileURLToPath, pathToFileURL } from 'url';
import { DiagnosticSeverity } from 'vscode-languageserver';
import { ProjectParser } from '../projectParser';
import { OIDiagnostic } from '../vhdlLinter';
import { getCodeClimate, prettyPrintMessages } from './cliUtil';
import { lintFolder } from './lintFolder';

export async function cli(argv: string[]) {
  const integerParam = (v: string) => {
    const n = Number(v);
    if (Number.isInteger(n)) {
      return n;
    }
    throw new InvalidArgumentError(`Not an integer`);
  };
  const folderParam = (v: string) => {
    const path = isAbsolute(v) ? v : join(cwd(), v);
    if (existsSync(path) === false) {
      throw new InvalidArgumentError("Does not exist");
    }
    if (lstatSync(path).isDirectory() === false) {
      throw new InvalidArgumentError("Is not a directory");
    }
    return pathToFileURL(path);
  };
  const program = new Command('vhdl-linter')
    .description('A typescript based linter for vhdl')
    .argument('<folder>', 'The folder to lint', folderParam)
    .option('-j, --output-json', 'Output message in json compatible with Code Climate Engine Specification (Gitlab & Github compatible)', false)
    .option('-e, --exclude <pattern...>', 'Exclude pattern for linting. Use the ignore setting from the `vhdl-linter.yml` file to not parse files at all.', [])
    .option('-i, --max-info <count>', 'Number of information messages to trigger a nonzero exit code', integerParam, -1)
    .option('-w, --max-warnings <count>', 'Number of information messages to trigger a nonzero exit code', integerParam, -1)
    .parse(argv);
  const options = program.opts();
  const start = new Date().getTime();
  const folder = program.processedArgs[0];
  const projectParser = await ProjectParser.create([folder]);
  const filesWithMessages = await lintFolder(folder, false, options.outputJson === false, options.exclude, projectParser);
  await projectParser.stop();
  const allMessages = filesWithMessages.flatMap(m => m.messages);
  const warningCount = allMessages.filter(m => m?.severity === DiagnosticSeverity.Warning).length;
  const infoCount = allMessages.filter(m => m?.severity === DiagnosticSeverity.Information).length;
  // default is error
  function isError(m: OIDiagnostic) {
    return m?.severity !== DiagnosticSeverity.Warning && m?.severity !== DiagnosticSeverity.Information;
  }
  const errorCount = allMessages.filter(isError).length;
  if (options.outputJson) {
    const rootPath = fileURLToPath(projectParser.workspaces[0]!);
    console.log(JSON.stringify(getCodeClimate(filesWithMessages, rootPath), undefined, 2));
    const filesWithErrors = filesWithMessages.map(fileWithMessages => ({ ...fileWithMessages, messages: fileWithMessages.messages.filter(isError) })).filter(fileWithMessages => fileWithMessages.messages.length > 0);
    console.error(prettyPrintMessages(rootPath, filesWithErrors));
  } else {
    const timeTaken = new Date().getTime() - start;
    console.log(`Linted in ${(timeTaken / 1000).toFixed(2)}s:`);
    console.log(`${errorCount} error(s), ${warningCount} warning(s), ${infoCount} info(s)`);
  }
  if (errorCount > 0) {
    return 1;
  } else if (warningCount > options.maxWarnings && options.maxWarnings !== -1) {
    return 1;
  } else if (infoCount > options.maxInfo && options.maxInfo !== -1) {
    return 1;
  }
  return 0;
}
