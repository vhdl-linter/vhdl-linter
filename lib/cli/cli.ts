#!/usr/bin/env node
import { cli } from './cliExec';

(async () => {
  const returnCode = await cli(process.argv);
  process.exit(returnCode);
})().catch(err => console.error(err));
