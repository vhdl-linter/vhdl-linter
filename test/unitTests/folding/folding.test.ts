import { afterAll, beforeAll, expect, test } from '@jest/globals';
import { readdirSync } from 'fs';
import { join } from 'path';
import { pathToFileURL } from 'url';
import { foldingHandler } from '../../../lib/languageFeatures/folding';
import { ProjectParser } from '../../../lib/projectParser';
import { VhdlLinter } from '../../../lib/vhdlLinter';
import { readFileSyncNorm } from "../../../lib/cli/readFileSyncNorm";


let projectParser: ProjectParser;
beforeAll(async () => {
  projectParser = await ProjectParser.create([pathToFileURL(__dirname)]);
});
afterAll(async () => {
  await projectParser.stop();
});

test.each(
  readdirSync(__dirname).filter(v => v.endsWith('.vhd'))
)('Testing folding ranges of %s', async (fileName) => {
  const path = join(__dirname, fileName);
  const uri = pathToFileURL(path);
  const linter = new VhdlLinter(uri, readFileSyncNorm(uri, { encoding: 'utf8' }), projectParser, await projectParser.getDocumentSettings(uri));
  const symbols = foldingHandler(linter);
  expect(symbols).toMatchSnapshot();
});