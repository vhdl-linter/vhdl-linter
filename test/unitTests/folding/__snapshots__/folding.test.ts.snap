// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing folding ranges of configuration.vhd 1`] = `
[
  {
    "endLine": 3,
    "startLine": 0,
  },
]
`;

exports[`Testing folding ranges of test_alias.vhd 1`] = `
[
  {
    "endLine": 10,
    "startLine": 4,
  },
  {
    "endLine": 2,
    "kind": "imports",
    "startLine": 0,
  },
  {
    "endLine": 7,
    "startLine": 5,
  },
]
`;

exports[`Testing folding ranges of test_block.vhd 1`] = `
[
  {
    "endLine": 2,
    "kind": "imports",
    "startLine": 0,
  },
  {
    "endLine": 8,
    "startLine": 7,
  },
  {
    "endLine": 14,
    "startLine": 9,
  },
  {
    "endLine": 12,
    "startLine": 11,
  },
  {
    "endLine": 0,
    "kind": "imports",
    "startLine": Infinity,
  },
]
`;

exports[`Testing folding ranges of test_case_generate.vhd 1`] = `
[
  {
    "endLine": 2,
    "kind": "imports",
    "startLine": 0,
  },
  {
    "endLine": 8,
    "startLine": 7,
  },
  {
    "endLine": 17,
    "startLine": 9,
  },
  {
    "endLine": 16,
    "startLine": 11,
  },
  {
    "endLine": 13,
    "startLine": 12,
  },
  {
    "endLine": 0,
    "kind": "imports",
    "startLine": Infinity,
  },
  {
    "endLine": 15,
    "startLine": 14,
  },
  {
    "endLine": 0,
    "kind": "imports",
    "startLine": Infinity,
  },
]
`;

exports[`Testing folding ranges of test_entity.vhd 1`] = `
[
  {
    "endLine": 3,
    "startLine": 1,
  },
]
`;

exports[`Testing folding ranges of test_for_generate.vhd 1`] = `
[
  {
    "endLine": 2,
    "kind": "imports",
    "startLine": 0,
  },
  {
    "endLine": 8,
    "startLine": 7,
  },
  {
    "endLine": 14,
    "startLine": 9,
  },
  {
    "endLine": 12,
    "startLine": 11,
  },
  {
    "endLine": 0,
    "kind": "imports",
    "startLine": Infinity,
  },
]
`;

exports[`Testing folding ranges of test_if_else_generate.vhd 1`] = `
[
  {
    "endLine": 2,
    "kind": "imports",
    "startLine": 0,
  },
  {
    "endLine": 8,
    "startLine": 7,
  },
  {
    "endLine": 16,
    "startLine": 9,
  },
  {
    "endLine": 12,
    "startLine": 11,
  },
  {
    "endLine": 0,
    "kind": "imports",
    "startLine": Infinity,
  },
  {
    "endLine": 14,
    "startLine": 13,
  },
  {
    "endLine": 0,
    "kind": "imports",
    "startLine": Infinity,
  },
]
`;

exports[`Testing folding ranges of test_if_elsif_generate.vhd 1`] = `
[
  {
    "endLine": 2,
    "kind": "imports",
    "startLine": 0,
  },
  {
    "endLine": 8,
    "startLine": 7,
  },
  {
    "endLine": 15,
    "startLine": 9,
  },
  {
    "endLine": 12,
    "startLine": 11,
  },
  {
    "endLine": 0,
    "kind": "imports",
    "startLine": Infinity,
  },
  {
    "endLine": 14,
    "startLine": 13,
  },
  {
    "endLine": 0,
    "kind": "imports",
    "startLine": Infinity,
  },
]
`;

exports[`Testing folding ranges of test_inst.vhd 1`] = `
[
  {
    "endLine": 2,
    "kind": "imports",
    "startLine": 0,
  },
  {
    "endLine": 8,
    "startLine": 7,
  },
  {
    "endLine": 16,
    "startLine": 9,
  },
  {
    "endLine": 14,
    "startLine": 11,
  },
  {
    "endLine": 14,
    "startLine": 12,
  },
]
`;

exports[`Testing folding ranges of test_process.vhd 1`] = `
[
  {
    "endLine": 2,
    "kind": "imports",
    "startLine": 0,
  },
  {
    "endLine": 7,
    "startLine": 5,
  },
  {
    "endLine": 11,
    "startLine": 10,
  },
  {
    "endLine": 20,
    "startLine": 12,
  },
  {
    "endLine": 14,
    "startLine": 14,
  },
  {
    "endLine": 18,
    "startLine": 15,
  },
]
`;

exports[`Testing folding ranges of test_protected.vhd 1`] = `
[
  {
    "endLine": 9,
    "startLine": 4,
  },
  {
    "endLine": 2,
    "kind": "imports",
    "startLine": 0,
  },
  {
    "endLine": 8,
    "startLine": 5,
  },
  {
    "endLine": 9,
    "startLine": 5,
  },
  {
    "endLine": 7,
    "startLine": 7,
  },
  {
    "endLine": 54,
    "startLine": 12,
  },
  {
    "endLine": 16,
    "startLine": 15,
  },
  {
    "endLine": 21,
    "startLine": 17,
  },
  {
    "endLine": 15,
    "startLine": 15,
  },
  {
    "endLine": 19,
    "startLine": 18,
  },
  {
    "endLine": 53,
    "startLine": 25,
  },
  {
    "endLine": 54,
    "startLine": 25,
  },
  {
    "endLine": 29,
    "startLine": 29,
  },
  {
    "endLine": 31,
    "startLine": 30,
  },
  {
    "endLine": 34,
    "startLine": 34,
  },
  {
    "endLine": 44,
    "startLine": 35,
  },
  {
    "endLine": 34,
    "startLine": 34,
  },
  {
    "endLine": 43,
    "startLine": 36,
  },
  {
    "endLine": 42,
    "startLine": 37,
  },
  {
    "endLine": 39,
    "startLine": 38,
  },
  {
    "endLine": 41,
    "startLine": 40,
  },
  {
    "endLine": 48,
    "startLine": 47,
  },
  {
    "endLine": 51,
    "startLine": 49,
  },
]
`;

exports[`Testing folding ranges of test_record.vhd 1`] = `
[
  {
    "endLine": 7,
    "startLine": 4,
  },
  {
    "endLine": 2,
    "kind": "imports",
    "startLine": 0,
  },
  {
    "endLine": 7,
    "startLine": 5,
  },
]
`;

exports[`Testing folding ranges of test_stateMachine.vhd 1`] = `
[
  {
    "endLine": 2,
    "kind": "imports",
    "startLine": 0,
  },
  {
    "endLine": 9,
    "startLine": 7,
  },
  {
    "endLine": 25,
    "startLine": 10,
  },
  {
    "endLine": 12,
    "startLine": 12,
  },
  {
    "endLine": 23,
    "startLine": 13,
  },
  {
    "endLine": 23,
    "startLine": 14,
  },
  {
    "endLine": 16,
    "startLine": 15,
  },
  {
    "endLine": 18,
    "startLine": 17,
  },
  {
    "endLine": 20,
    "startLine": 19,
  },
  {
    "endLine": 22,
    "startLine": 21,
  },
]
`;

exports[`Testing folding ranges of test_stateMachineCase.vhd 1`] = `
[
  {
    "endLine": 2,
    "kind": "imports",
    "startLine": 0,
  },
  {
    "endLine": 10,
    "startLine": 7,
  },
  {
    "endLine": 31,
    "startLine": 11,
  },
  {
    "endLine": 13,
    "startLine": 13,
  },
  {
    "endLine": 29,
    "startLine": 14,
  },
  {
    "endLine": 29,
    "startLine": 15,
  },
  {
    "endLine": 22,
    "startLine": 16,
  },
  {
    "endLine": 22,
    "startLine": 17,
  },
  {
    "endLine": 19,
    "startLine": 18,
  },
  {
    "endLine": 21,
    "startLine": 20,
  },
  {
    "endLine": 24,
    "startLine": 23,
  },
  {
    "endLine": 26,
    "startLine": 25,
  },
  {
    "endLine": 28,
    "startLine": 27,
  },
]
`;

exports[`Testing folding ranges of test_stateMachineElse.vhd 1`] = `
[
  {
    "endLine": 2,
    "kind": "imports",
    "startLine": 0,
  },
  {
    "endLine": 9,
    "startLine": 7,
  },
  {
    "endLine": 28,
    "startLine": 10,
  },
  {
    "endLine": 12,
    "startLine": 12,
  },
  {
    "endLine": 26,
    "startLine": 13,
  },
  {
    "endLine": 14,
    "startLine": 14,
  },
  {
    "endLine": 25,
    "startLine": 15,
  },
  {
    "endLine": 25,
    "startLine": 16,
  },
  {
    "endLine": 18,
    "startLine": 17,
  },
  {
    "endLine": 20,
    "startLine": 19,
  },
  {
    "endLine": 22,
    "startLine": 21,
  },
  {
    "endLine": 24,
    "startLine": 23,
  },
]
`;

exports[`Testing folding ranges of test_stateMachineElsif.vhd 1`] = `
[
  {
    "endLine": 2,
    "kind": "imports",
    "startLine": 0,
  },
  {
    "endLine": 9,
    "startLine": 7,
  },
  {
    "endLine": 28,
    "startLine": 10,
  },
  {
    "endLine": 12,
    "startLine": 12,
  },
  {
    "endLine": 26,
    "startLine": 13,
  },
  {
    "endLine": 14,
    "startLine": 14,
  },
  {
    "endLine": 25,
    "startLine": 15,
  },
  {
    "endLine": 25,
    "startLine": 16,
  },
  {
    "endLine": 18,
    "startLine": 17,
  },
  {
    "endLine": 20,
    "startLine": 19,
  },
  {
    "endLine": 22,
    "startLine": 21,
  },
  {
    "endLine": 24,
    "startLine": 23,
  },
]
`;

exports[`Testing folding ranges of test_stateMachineFor.vhd 1`] = `
[
  {
    "endLine": 2,
    "kind": "imports",
    "startLine": 0,
  },
  {
    "endLine": 9,
    "startLine": 7,
  },
  {
    "endLine": 27,
    "startLine": 10,
  },
  {
    "endLine": 12,
    "startLine": 12,
  },
  {
    "endLine": 25,
    "startLine": 13,
  },
  {
    "endLine": 24,
    "startLine": 14,
  },
  {
    "endLine": 24,
    "startLine": 15,
  },
  {
    "endLine": 17,
    "startLine": 16,
  },
  {
    "endLine": 19,
    "startLine": 18,
  },
  {
    "endLine": 21,
    "startLine": 20,
  },
  {
    "endLine": 23,
    "startLine": 22,
  },
]
`;

exports[`Testing folding ranges of test_stateMachineIf.vhd 1`] = `
[
  {
    "endLine": 2,
    "kind": "imports",
    "startLine": 0,
  },
  {
    "endLine": 9,
    "startLine": 7,
  },
  {
    "endLine": 27,
    "startLine": 10,
  },
  {
    "endLine": 12,
    "startLine": 12,
  },
  {
    "endLine": 25,
    "startLine": 13,
  },
  {
    "endLine": 24,
    "startLine": 14,
  },
  {
    "endLine": 24,
    "startLine": 15,
  },
  {
    "endLine": 17,
    "startLine": 16,
  },
  {
    "endLine": 19,
    "startLine": 18,
  },
  {
    "endLine": 21,
    "startLine": 20,
  },
  {
    "endLine": 23,
    "startLine": 22,
  },
]
`;

exports[`Testing folding ranges of test_subprogram.vhd 1`] = `
[
  {
    "endLine": 5,
    "startLine": 4,
  },
  {
    "endLine": 2,
    "kind": "imports",
    "startLine": 0,
  },
  {
    "endLine": 12,
    "startLine": 8,
  },
  {
    "endLine": 9,
    "startLine": 9,
  },
  {
    "endLine": 11,
    "startLine": 10,
  },
]
`;
